# Coefficient B12

SS final-interval case: `i=N, j=N`, with $n=N-1$. This is a new derivation in the author’s coefficient layout, using **p** for normalized power. It replaces the corresponding final-interval placement; earlier intervals remain unchanged. SN transposes NS.

## Definition

Set $\chi_1=\chi_{N-1}$, $\chi_2=\chi_N$ and $\Delta\chi=\chi_2-\chi_1$.
The density hats are $L(\chi)=(\chi_2-\chi)/\Delta\chi$ and $R(\chi)=(\chi-\chi_1)/\Delta\chi$.
For $\chi_1\leq\chi\leq\chi_2$, the source lower limit is the evaluation point:

$$F(\chi)=\int_\chi^{\chi_2}\frac{\chi_2-u}{\Delta\chi}\frac{u-\chi}{u}\,du,
\qquad H(\chi)=\int_\chi^{\chi_2}\frac{u-\chi_1}{\Delta\chi}\frac{u-\chi}{u}\,du.$$

There is no source half beyond $\chi_N$. Below $\chi_1$, the rising source instead has its full support $[\chi_1,\chi_2]$. At $\chi_1>0$ the value and first derivative match; both partial source integrals vanish at $\chi_2$.


## Change of variables

$$t=\frac{\chi}{\chi_2},\quad a=1-\frac{\chi_1}{\chi_2},\quad
p=1-\frac{P_1}{P_2},\quad z=1-\frac{1+z_1}{1+z_2}.$$

For an ordinary interval $0<a<1$, $1-a\leq t\leq1$, and $p,z$ are real. The normalized power notation presumes $P_2\neq0$; an endpoint-linear equivalent below removes that restriction. $P_1,P_2$ may be signed effective component power.

$$Q(t)=1-\frac{p}{a}(1-t),\qquad Z(t)=1-\frac{z}{a}(1-t).$$
The integrated source functions are $F(\chi)=\chi_2F_a(t)$ and $H(\chi)=\chi_2H_a(t)$:

$$F_a(t)=\frac{1-t^2}{2a}+\frac{t\log t}{a},$$
$$H_a(t)=\frac{(1-t)^2}{2a}-\frac{(1-a)(1-t)}a-\frac{(1-a)t\log t}a.$$
The continuous endpoint value is $t\log t\to0$ as $t\to0^+$. Physical lensing amplitudes are applied by the component’s existing factor/effective-power convention, exactly once; these radial basis coefficients introduce no additional amplitude.


In [1]:
import sympy as sp
from IPython.display import display

t, a, p, z, b = sp.symbols('t a p z b', real=True)
Q = 1-p*(1-t)/a
Z = 1-z*(1-t)/a
F = ((1-t*t)/2+t*sp.log(t))/a
H = ((1-t)**2/2-(1-a)*(1-t)-(1-a)*t*sp.log(t))/a
u = sp.symbols('u', positive=True)
# Moving-limit source integration, expressed in the same normalized units.
F_PRIMITIVE = sp.integrate((1-u)*(u-t)/(a*u), u)
H_PRIMITIVE = sp.integrate((u-1+a)*(u-t)/(a*u), u)
assert sp.simplify(F_PRIMITIVE.subs(u,1)-F_PRIMITIVE.subs(u,t)-F) == 0
assert sp.simplify(H_PRIMITIVE.subs(u,1)-H_PRIMITIVE.subs(u,t)-H) == 0
display(F, H)


(-t**2/2 + t*log(t) + 1/2)/a

(-t*(1 - a)*log(t) - (1 - a)*(1 - t) + (1 - t)**2/2)/a

## Ordinary coefficient $I_{12}$

$$I_{12}=\int_{1-a}^1 \frac{\left(1 - \frac{p \left(1 - t\right)}{a}\right) \left(1 - \frac{z \left(1 - t\right)}{a}\right)^{2} \left(- t \left(1 - a\right) \log{\left(t \right)} - \left(1 - a\right) \left(1 - t\right) + \frac{\left(1 - t\right)^{2}}{2}\right)^{2}}{a^{2}}\,dt.$$

The dimensional coefficient is $B_{12}=\chi_2^3 P_2(1+z_2)^2 I_{12}$. Expand into powers of $t$ and $\log t$, then integrate each moment. For $k\neq-1$, the primitive of $t^k(\log t)^r$ is obtained by integration by parts; for $k=-1$, use $(\log t)^{r+1}/(r+1)$.

In [2]:
def integrate_moments(expression, lower):
    """Integrate every t**n log(t)**r term exactly on [lower, 1]."""
    result = 0
    for term in sp.Add.make_args(sp.expand(expression)):
        powers = term.as_powers_dict()
        n = int(powers.get(t, 0))
        r = int(powers.get(sp.log(t), 0))
        coefficient = term / t**n / sp.log(t)**r
        if n == -1:
            result -= coefficient*sp.log(lower)**(r+1)/sp.Integer(r+1)
        else:
            k = sp.Integer(n+1)
            primitive = t**k*sum((-1)**j*sp.factorial(r)/sp.factorial(r-j)
                                *sp.log(t)**(r-j)/k**(j+1) for j in range(r+1))
            boundary = 0 if lower == 0 else primitive.subs(t, lower)
            result += coefficient*(primitive.subs(t, 1)-boundary)
    return sp.factor(result)

INTEGRAND = H**2*Q*Z**2
EXPANDED = sp.expand(INTEGRAND)
I12 = integrate_moments(EXPANDED, 1-a)
display(EXPANDED, I12)


t**2*log(t)**2 - 2*t**2*log(t) + t**2 + 2*t*log(t) - 2*t + 1 + p*t**3*log(t)**2/a - 2*p*t**3*log(t)/a + p*t**3/a - p*t**2*log(t)**2/a + 4*p*t**2*log(t)/a - 3*p*t**2/a - 2*p*t*log(t)/a + 3*p*t/a - p/a + 2*t**3*z*log(t)**2/a - 4*t**3*z*log(t)/a + 2*t**3*z/a + t**3*log(t)/a - t**3/a - 2*t**2*z*log(t)**2/a + 8*t**2*z*log(t)/a - 6*t**2*z/a - 2*t**2*log(t)**2/a + 2*t**2*log(t)/a + t**2/a - 4*t*z*log(t)/a + 6*t*z/a - 3*t*log(t)/a + t/a - 2*z/a - 1/a + 2*p*t**4*z*log(t)**2/a**2 - 4*p*t**4*z*log(t)/a**2 + 2*p*t**4*z/a**2 + p*t**4*log(t)/a**2 - p*t**4/a**2 - 4*p*t**3*z*log(t)**2/a**2 + 12*p*t**3*z*log(t)/a**2 - 8*p*t**3*z/a**2 - 2*p*t**3*log(t)**2/a**2 + p*t**3*log(t)/a**2 + 2*p*t**3/a**2 + 2*p*t**2*z*log(t)**2/a**2 - 12*p*t**2*z*log(t)/a**2 + 12*p*t**2*z/a**2 + 2*p*t**2*log(t)**2/a**2 - 5*p*t**2*log(t)/a**2 + 4*p*t*z*log(t)/a**2 - 8*p*t*z/a**2 + 3*p*t*log(t)/a**2 - 2*p*t/a**2 + 2*p*z/a**2 + p/a**2 + t**4*z**2*log(t)**2/a**2 - 2*t**4*z**2*log(t)/a**2 + t**4*z**2/a**2 + 2*t**4*z*log(t)/a**2 - 2*t

-(882000*a**8*p*z**2*log(1 - a)**2 - 2814000*a**8*p*z**2*log(1 - a) + 2254375*a**8*p*z**2 - 2116800*a**8*p*z*log(1 - a)**2 + 6844320*a**8*p*z*log(1 - a) - 5568864*a**8*p*z + 1323000*a**8*p*log(1 - a)**2 - 4365900*a**8*p*log(1 - a) + 3640455*a**8*p - 1058400*a**8*z**2*log(1 - a)**2 + 3422160*a**8*z**2*log(1 - a) - 2784432*a**8*z**2 + 2646000*a**8*z*log(1 - a)**2 - 8731800*a**8*z*log(1 - a) + 7280910*a**8*z - 1764000*a**8*log(1 - a)**2 + 6027000*a**8*log(1 - a) - 5250350*a**8 - 3880800*a**7*p*z**2*log(1 - a)**2 + 8364720*a**7*p*z**2*log(1 - a) - 3445544*a**7*p*z**2 + 9525600*a**7*p*z*log(1 - a)**2 - 20921040*a**7*p*z*log(1 - a) + 8844108*a**7*p*z - 6174000*a**7*p*log(1 - a)**2 + 13994400*a**7*p*log(1 - a) - 6199480*a**7*p + 4762800*a**7*z**2*log(1 - a)**2 - 10460520*a**7*z**2*log(1 - a) + 4422054*a**7*z**2 - 12348000*a**7*z*log(1 - a)**2 + 27988800*a**7*z*log(1 - a) - 12398960*a**7*z + 8820000*a**7*log(1 - a)**2 - 21315000*a**7*log(1 - a) + 10522750*a**7 + 6438600*a**6*p*z**2*log(1 - a)*

### Result

$$I_{12}=- \frac{882000 a^{8} p z^{2} \log{\left(1 - a \right)}^{2} - 2814000 a^{8} p z^{2} \log{\left(1 - a \right)} + 2254375 a^{8} p z^{2} - 2116800 a^{8} p z \log{\left(1 - a \right)}^{2} + 6844320 a^{8} p z \log{\left(1 - a \right)} - 5568864 a^{8} p z + 1323000 a^{8} p \log{\left(1 - a \right)}^{2} - 4365900 a^{8} p \log{\left(1 - a \right)} + 3640455 a^{8} p - 1058400 a^{8} z^{2} \log{\left(1 - a \right)}^{2} + 3422160 a^{8} z^{2} \log{\left(1 - a \right)} - 2784432 a^{8} z^{2} + 2646000 a^{8} z \log{\left(1 - a \right)}^{2} - 8731800 a^{8} z \log{\left(1 - a \right)} + 7280910 a^{8} z - 1764000 a^{8} \log{\left(1 - a \right)}^{2} + 6027000 a^{8} \log{\left(1 - a \right)} - 5250350 a^{8} - 3880800 a^{7} p z^{2} \log{\left(1 - a \right)}^{2} + 8364720 a^{7} p z^{2} \log{\left(1 - a \right)} - 3445544 a^{7} p z^{2} + 9525600 a^{7} p z \log{\left(1 - a \right)}^{2} - 20921040 a^{7} p z \log{\left(1 - a \right)} + 8844108 a^{7} p z - 6174000 a^{7} p \log{\left(1 - a \right)}^{2} + 13994400 a^{7} p \log{\left(1 - a \right)} - 6199480 a^{7} p + 4762800 a^{7} z^{2} \log{\left(1 - a \right)}^{2} - 10460520 a^{7} z^{2} \log{\left(1 - a \right)} + 4422054 a^{7} z^{2} - 12348000 a^{7} z \log{\left(1 - a \right)}^{2} + 27988800 a^{7} z \log{\left(1 - a \right)} - 12398960 a^{7} z + 8820000 a^{7} \log{\left(1 - a \right)}^{2} - 21315000 a^{7} \log{\left(1 - a \right)} + 10522750 a^{7} + 6438600 a^{6} p z^{2} \log{\left(1 - a \right)}^{2} - 8205540 a^{6} p z^{2} \log{\left(1 - a \right)} + 1314733 a^{6} p z^{2} - 16228800 a^{6} p z \log{\left(1 - a \right)}^{2} + 21191520 a^{6} p z \log{\left(1 - a \right)} - 3493504 a^{6} p z + 11025000 a^{6} p \log{\left(1 - a \right)}^{2} - 15067500 a^{6} p \log{\left(1 - a \right)} + 2630075 a^{6} p - 8114400 a^{6} z^{2} \log{\left(1 - a \right)}^{2} + 10595760 a^{6} z^{2} \log{\left(1 - a \right)} - 1746752 a^{6} z^{2} + 22050000 a^{6} z \log{\left(1 - a \right)}^{2} - 30135000 a^{6} z \log{\left(1 - a \right)} + 5260150 a^{6} z - 17640000 a^{6} \log{\left(1 - a \right)}^{2} + 27636000 a^{6} \log{\left(1 - a \right)} - 6786500 a^{6} - 4762800 a^{5} p z^{2} \log{\left(1 - a \right)}^{2} + 2640120 a^{5} p z^{2} \log{\left(1 - a \right)} - 12474 a^{5} p z^{2} + 12348000 a^{5} p z \log{\left(1 - a \right)}^{2} - 6997200 a^{5} p z \log{\left(1 - a \right)} + 42140 a^{5} p z - 8820000 a^{5} p \log{\left(1 - a \right)}^{2} + 4998000 a^{5} p \log{\left(1 - a \right)} + 274400 a^{5} p + 6174000 a^{5} z^{2} \log{\left(1 - a \right)}^{2} - 3498600 a^{5} z^{2} \log{\left(1 - a \right)} + 21070 a^{5} z^{2} - 17640000 a^{5} z \log{\left(1 - a \right)}^{2} + 9996000 a^{5} z \log{\left(1 - a \right)} + 548800 a^{5} z + 17640000 a^{5} \log{\left(1 - a \right)}^{2} - 15288000 a^{5} \log{\left(1 - a \right)} + 1102500 a^{5} + 1323000 a^{4} p z^{2} \log{\left(1 - a \right)}^{2} - 14700 a^{4} p z^{2} \log{\left(1 - a \right)} - 18585 a^{4} p z^{2} - 3528000 a^{4} p z \log{\left(1 - a \right)}^{2} + 235200 a^{4} p z \log{\left(1 - a \right)} - 125440 a^{4} p z + 2205000 a^{4} p \log{\left(1 - a \right)}^{2} + 955500 a^{4} p \log{\left(1 - a \right)} - 22050 a^{4} p - 1764000 a^{4} z^{2} \log{\left(1 - a \right)}^{2} + 117600 a^{4} z^{2} \log{\left(1 - a \right)} - 62720 a^{4} z^{2} + 4410000 a^{4} z \log{\left(1 - a \right)}^{2} + 1911000 a^{4} z \log{\left(1 - a \right)} - 44100 a^{4} z - 8820000 a^{4} \log{\left(1 - a \right)}^{2} + 2793000 a^{4} \log{\left(1 - a \right)} + 147000 a^{4} - 58800 a^{3} p z^{2} \log{\left(1 - a \right)} + 5460 a^{3} p z^{2} + 352800 a^{3} p z \log{\left(1 - a \right)}^{2} - 552720 a^{3} p z \log{\left(1 - a \right)} - 229320 a^{3} p z + 882000 a^{3} p \log{\left(1 - a \right)}^{2} - 411600 a^{3} p \log{\left(1 - a \right)} - 102900 a^{3} p + 176400 a^{3} z^{2} \log{\left(1 - a \right)}^{2} - 276360 a^{3} z^{2} \log{\left(1 - a \right)} - 114660 a^{3} z^{2} + 1764000 a^{3} z \log{\left(1 - a \right)}^{2} - 823200 a^{3} z \log{\left(1 - a \right)} - 205800 a^{3} z + 1764000 a^{3} \log{\left(1 - a \right)}^{2} + 147000 a^{3} \log{\left(1 - a \right)} - 88200 a^{2} p z^{2} \log{\left(1 - a \right)}^{2} + 79380 a^{2} p z^{2} \log{\left(1 - a \right)} + 126630 a^{2} p z^{2} - 705600 a^{2} p z \log{\left(1 - a \right)}^{2} + 47040 a^{2} p z \log{\left(1 - a \right)} + 152880 a^{2} p z - 441000 a^{2} p \log{\left(1 - a \right)}^{2} - 102900 a^{2} p \log{\left(1 - a \right)} - 352800 a^{2} z^{2} \log{\left(1 - a \right)}^{2} + 23520 a^{2} z^{2} \log{\left(1 - a \right)} + 76440 a^{2} z^{2} - 882000 a^{2} z \log{\left(1 - a \right)}^{2} - 205800 a^{2} z \log{\left(1 - a \right)} + 176400 a p z^{2} \log{\left(1 - a \right)}^{2} + 68040 a p z^{2} \log{\left(1 - a \right)} - 59220 a p z^{2} + 352800 a p z \log{\left(1 - a \right)}^{2} + 152880 a p z \log{\left(1 - a \right)} + 176400 a z^{2} \log{\left(1 - a \right)}^{2} + 76440 a z^{2} \log{\left(1 - a \right)} - 88200 p z^{2} \log{\left(1 - a \right)}^{2} - 59220 p z^{2} \log{\left(1 - a \right)}}{5292000 a^{5}}.$$

## Observer coefficient $J_{12}$

Set $a=1$, but replace the power with $Q(t)=t^3$ **before** integration. Substituting $p=1$ into $I_{12}$ would preserve linear power and is not this observer convention. The dimensional prefactor is the same.

In [3]:
OBSERVER_INTEGRAND = sp.cancel(INTEGRAND/Q).subs(a,1)*t**3
J12 = integrate_moments(OBSERVER_INTEGRAND, 0)
display(OBSERVER_INTEGRAND, J12)


t**3*(t**6*z**2/4 - 3*t**5*z**2/2 + t**5*z/2 + 15*t**4*z**2/4 - 5*t**4*z/2 + t**4/4 - 5*t**3*z**2 + 5*t**3*z - t**3 + 15*t**2*z**2/4 - 5*t**2*z + 3*t**2/2 - 3*t*z**2/2 + 5*t*z/2 - t + z**2/4 - z/2 + 1/4)

(3*z**2 - 10*z + 9)/10080

### Result

$$J_{12}=\frac{z^{2}}{3360} - \frac{z}{1008} + \frac{1}{1120}.$$

## Correspondence with the existing interior source

The falling half after the source node has width $b=(\chi_{n+2}-\chi_{n+1})/\chi_{n+1}$. Its full-support contribution is $D(t,b)=b/2+t[1-(1+b)\log(1+b)/b]$. Therefore $D\to0$ as $b\to0^+$, and the interior source $H+D$ becomes $H$. The following also verifies the integrated result against the original SS B05 text export, rather than directly substituting into a singular $1/b$ expression.

In [4]:
import re
from pathlib import Path

D = b/2+t*(1-(1+b)*sp.log(1+b)/b)
assert sp.limit(D,b,0,dir='+') == 0
INTERIOR_INTEGRAND = (H+D)**2*Q*Z**2
assert sp.simplify(sp.limit(INTERIOR_INTEGRAND,b,0,dir='+')-INTEGRAND) == 0
FILENAME = 'Coefficient_B05.txt'
HERE = Path.cwd()
for CANDIDATE in (HERE, HERE/'notebooks'/'derivation'/'SS'):
    if (CANDIDATE/FILENAME).is_file():
        HERE = CANDIDATE
        break
ORIGINAL = (HERE/FILENAME).read_text()
PARTS = re.split(r'\b[IJ]\d+\s*=', ORIGINAL)[1:]
for TEXT, TARGET in zip(PARTS, (I12,J12), strict=True):
    EXPR = sp.sympify(TEXT.replace('numpy.log','log').replace('\n',' '), locals={'a':a,'b':b,'p':p,'z':z,'log':sp.log})
    # Denominator degree is <=2: terms through b^4 preserve its finite limit.
    REGULAR = sp.cancel(EXPR.subs(sp.log(1+b), b-b**2/2+b**3/3-b**4/4))
    assert sp.factor(REGULAR.subs(b,0)-TARGET) == 0
print('Both I and J agree with the original interior-source limit.')


Both I and J agree with the original interior-source limit.


## Stable evaluation and zero endpoint powers

Use $s=(1-t)/a\in[0,1]$ and $d_m=1/[(m+2)(m+3)]$. Define
$$f(v)=\sum_{m=0}^{\infty}d_m v^m,\qquad
g(v)=\sum_{m=0}^{\infty}\left(\frac12-\frac1{m+3}\right)v^m.$$
Then
$$F_a=a^2s^3f(as),\quad H_a=\frac{a^2s^2}2-(1-a)F_a,$$
$$\frac{F_a}{t}=a^2s^3g(as),\quad
\frac{H_a}{t}=\frac{a^2s^2}{2(1-as)}-(1-a)\frac{F_a}{t}.$$
For SS, $f(v)^2=\sum e_m v^m$, with $e_m=\sum_{j=0}^m d_jd_{m-j}$; multiply the two source expressions before integrating.

All remaining factors are polynomials in $s$:
$$P(s)=P_2+(P_1-P_2)s,\qquad (1+z)(s)=(1+z_2)+(z_1-z_2)s.$$
Integrate each resulting monomial using $\int_0^1s^qds=1/(q+1)$. This gives analytic series of **integrated moments**, with no production quadrature. The NS scale is $\chi_2a^3$, and the SS scale is $\chi_2^3a^5$ when endpoint powers/redshifts are included in the moments.

Production uses 16 terms for $a\le1/16$, 32 for $a\le1/4$, 64 for $a\le1/2$, and 160 for $a\le3/4$. The geometric tail parameter $a^K$ is below $6\times10^{-20}$ in each regime. The coefficients and moment factors are bounded, with at most a factor $1/(1-a)\le4$ from the geometric tail. Broader intervals use the displayed elementary closed form; the observer uses its separate $J$ expression. Floating-point cancellation from an intentionally sign-changing final spectrum is assessed with absolute error too.

The direct endpoint-linear moments avoid $P_1/P_2$. For the closed form, $I=A+Bp$ implies
$$P_2I=-BP_1+(A+B)P_2.$$
This equivalence handles either endpoint being zero without changing a signed contribution or introducing an epsilon.


In [5]:
LEFT_WEIGHT = sp.factor(-sp.diff(I12,p))
RIGHT_WEIGHT = sp.factor(I12.subs(p,0)-LEFT_WEIGHT)
assert sp.simplify(I12-(LEFT_WEIGHT*(1-p)+RIGHT_WEIGHT)) == 0
display(LEFT_WEIGHT, RIGHT_WEIGHT)


(882000*a**8*z**2*log(1 - a)**2 - 2814000*a**8*z**2*log(1 - a) + 2254375*a**8*z**2 - 2116800*a**8*z*log(1 - a)**2 + 6844320*a**8*z*log(1 - a) - 5568864*a**8*z + 1323000*a**8*log(1 - a)**2 - 4365900*a**8*log(1 - a) + 3640455*a**8 - 3880800*a**7*z**2*log(1 - a)**2 + 8364720*a**7*z**2*log(1 - a) - 3445544*a**7*z**2 + 9525600*a**7*z*log(1 - a)**2 - 20921040*a**7*z*log(1 - a) + 8844108*a**7*z - 6174000*a**7*log(1 - a)**2 + 13994400*a**7*log(1 - a) - 6199480*a**7 + 6438600*a**6*z**2*log(1 - a)**2 - 8205540*a**6*z**2*log(1 - a) + 1314733*a**6*z**2 - 16228800*a**6*z*log(1 - a)**2 + 21191520*a**6*z*log(1 - a) - 3493504*a**6*z + 11025000*a**6*log(1 - a)**2 - 15067500*a**6*log(1 - a) + 2630075*a**6 - 4762800*a**5*z**2*log(1 - a)**2 + 2640120*a**5*z**2*log(1 - a) - 12474*a**5*z**2 + 12348000*a**5*z*log(1 - a)**2 - 6997200*a**5*z*log(1 - a) + 42140*a**5*z - 8820000*a**5*log(1 - a)**2 + 4998000*a**5*log(1 - a) + 274400*a**5 + 1323000*a**4*z**2*log(1 - a)**2 - 14700*a**4*z**2*log(1 - a) - 18585*a**4*

(176400*a**8*z**2*log(1 - a)**2 - 608160*a**8*z**2*log(1 - a) + 530057*a**8*z**2 - 529200*a**8*z*log(1 - a)**2 + 1887480*a**8*z*log(1 - a) - 1712046*a**8*z + 441000*a**8*log(1 - a)**2 - 1661100*a**8*log(1 - a) + 1609895*a**8 - 882000*a**7*z**2*log(1 - a)**2 + 2095800*a**7*z**2*log(1 - a) - 976510*a**7*z**2 + 2822400*a**7*z*log(1 - a)**2 - 7067760*a**7*z*log(1 - a) + 3554852*a**7*z - 2646000*a**7*log(1 - a)**2 + 7320600*a**7*log(1 - a) - 4323270*a**7 + 1675800*a**6*z**2*log(1 - a)**2 - 2390220*a**6*z**2*log(1 - a) + 432019*a**6*z**2 - 5821200*a**6*z*log(1 - a)**2 + 8943480*a**6*z*log(1 - a) - 1766646*a**6*z + 6615000*a**6*log(1 - a)**2 - 12568500*a**6*log(1 - a) + 4156425*a**6 - 1411200*a**5*z**2*log(1 - a)**2 + 858480*a**5*z**2*log(1 - a) - 8596*a**5*z**2 + 5292000*a**5*z*log(1 - a)**2 - 2998800*a**5*z*log(1 - a) - 590940*a**5*z - 8820000*a**5*log(1 - a)**2 + 10290000*a**5*log(1 - a) - 1376900*a**5 + 441000*a**4*z**2*log(1 - a)**2 - 102900*a**4*z**2*log(1 - a) + 81305*a**4*z**2 - 88200

## Export and validation

[Coefficient_B12.txt](Coefficient_B12.txt) contains the ordinary and observer expressions in the existing `I12 = ...`, `J12 = ...` format. [Coefficient_B12_Validation.ipynb](Coefficient_B12_Validation.ipynb) compares the compiled float64 coefficient against independent integration of the original hats and moving source limits. The original Mathematica derivations remain unchanged.